File: java/src/Subsets_78.java
import java.util.ArrayList;
import java.util.List;

public class Subsets_78 {

    public void recursion(List<List<Integer>> res, List<Integer> temp, int[] nums, int start) {
        res.add(new ArrayList<>(temp));
        if (temp.size() >= nums.length) return;
        for (int i = start; i < nums.length; i++) {
            temp.add(nums[i]);
            recursion(res, temp, nums, i + 1);
            temp.remove(temp.size() - 1);
        }
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        recursion(res, new ArrayList<>(), nums, 0);
        return res;
    }

    public static void main(String[] args) {
        System.out.println(new Subsets_78().subsets(new int[]{1,2,3}));
    }
}

File: go/classic/subsets_78.go
package classic

func subsets(nums []int) [][]int {
	res := make([][]int, 0)
	recursionSubsets(&res, nums, []int{}, 0)
	return res
}

func recursionSubsets(res *[][]int, nums []int, tmp []int, start int) {
	cp := make([]int, len(tmp))
	copy(cp, tmp)
	*res = append(*res, cp)
	for i := start; i < len(nums); i++ {
		tmp = append(tmp, nums[i])
		recursionSubsets(res, nums, tmp, i+1)
		tmp = tmp[:len(tmp)-1]
	}
}
